add invariants spec (#443)
apache/arrow-datafusion: docs/specification/invariants.md

* Logical field: Field with a relation name. Represented in this document by `LF(relation, name, type, nullable)`
* Projected plan: plan with projection as the root node.
* Logical schema: a vector of logical fields, used by logical plan.
* Physical schema: a vector of physical fields, used by both physical plan and Arrow record batch.

### Logical

#### Function

An object that knows its valid incoming logical fields and how to derive its
output logical field from its arguments' logical fields. A functions' output
field is itself a function of its input fields:

```
logical_field(lf1: LF, lf2: LF, ...) -> LF
```

Examples:

* `plus(a,b) -> LF(None, "{a} Plus {b}", d(a.type,b.type), a.nullable | b.nullable)` where d is the function mapping input types to output type (`get_supertype` in our current implementation).
* `length(a) -> LF(None, "length({a})", u32, a.nullable)`

#### Plan

A tree composed of other plans and functions (e.g. `Projection c1 + c2, c1 - c2 AS sum12; Scan c1 as u32, c2 as u64`)
that knows how to derive its schema.

Certain plans have a frozen schema (e.g. Scan), while others derive their
schema from their child nodes.

#### Column

> A type of logical node in a logical plan consists of field name and relation name.

Reviewer: ```suggestion
An identifier in a logical plan consists of field name and relation name.
```

### Physical

#### Function

An object that knows how to derive its physical field from its arguments'
physical fields, and also how to actually perform the computation on data. A
functions' output physical field is a function of its input physical fields:

```
physical_field(PF1, PF2, ...) -> PF
```

Examples:

* `plus(a,b) -> PF("{a} Plus {b}", d(a.type,b.type), a.nullable | b.nullable)` where d is a complex function (`get_supertype` in our current implementation) whose computation is for each element in the columns, sum the two entries together and return it in the same type as the smallest type of both columns.
* `length(&str) -> PF("length({a})", u32, a.nullable)` whose computation is "count number of bytes in the string".

#### Plan

A tree (e.g. `Projection c1 + c2, c1 - c2 AS sum12; Scan c1 as u32, c2 as u64`)
that knows how to derive its metadata and compute itself.

Note how the physical plane does not know how to derive field names: field

> names are solely a property of the logical plane, as they are not needed in the

Reviewer: I don't understand the assertion that field names aren't needed for physical plans. Field names are part of the `Schema` that is passed around in Physical plans and the names are used to line up inputs to PhysicalExpr

Author: @jorgecarleitao please correct me if I am wrong. I think the focus on this statement is on generation of field names. In the current code base, physical field names are generated using metadata from the logical plane, more specifically  from logical `Expr`s. The actual physical plan and physical expressions are not used to generated the field name.

> names are used to line up inputs to PhysicalExpr

This used to be true at the time when Jorge wrote it, but with https://github.com/apache/arrow-datafusion/pull/55, we will be changing it to line up inputs using index instead that. But regardless we make this change or not, it doesn't change the fact that physical filed names are only generated using metadata from the logical plane.

Reviewer 2: That is my understanding also: naming is in the logical plane; in the physical plane, people can use whatever they want for as long as the record batches respect the output schema declared on the corresponding logical nodes.

People can derive a debug format for the physical expressions, but the naming must be set beforehand, or we risk having a divergence between the names declared in the output schema of the logical node and the record batches returned by the physical node.

Reviewer: I think I misread the text to say the physical plans had no field names

only be derived based on a known set of input schemas (TableProvider).
Likewise, schemas of functions depend on the specific registry of functions
registered (e.g. does `my_op` return u32 or u64?). Thus, in this section, the
wording "same schema" is understood to mean "same schema under a given registry
of data sources and functions".

### (relation, name) tuples in logical fields and logical columns are unique

Every logical field's (relation, name) tuple in a logical schema MUST be unique.
Every logical column's (relation, name) tuple in a logical plan MUST be unique.

This invariant guarantees that `SELECT t1.id, t2.id FROM t1 JOIN t2...`
unambiguously selects the field `t1.id` and `t2.id` in a logical schema in the
logical plane.

#### Responsibility

It is the logical builder and optimizer's responsibility to guarantee this
invariant.

#### Validation

Builder and optimizer MUST error if this invariant is violated on any logical
node that creates a new schema (e.g. scan, projection, aggregation, join, etc.).

### Physical schema is consistent with data

The contents of every Array in every RecordBatch in every partition returned by

> a physical plan MUST be consistent with recordBatch's schema, in that every

Reviewer: ```suggestion
a physical plan MUST be consistent with RecordBatch's schema, in that every
```

Array in the RecordBatch must be downcastable to its corresponding type
declared in the RecordBatch.

#### Responsibility

Physical functions MUST guarantee this invariant. This is particularly
important in aggregate functions, whose aggregating type may be different from
the intermediary types during calculations (e.g. sum(i32) -> i64).

#### Validation

Since the validation of this invariant is computationally expensive, execution
contexts CAN validate this invariant. It is acceptable for physical nodes to
`panic!` if their input does not satisfy this invariant.

### Physical schema is consistent in physical functions

The schema of every Array returned by a physical function MUST match the

> DataType derived by the physical function itself.

Reviewer: ```suggestion
DataType reported by the physical function itself.
```

This ensures that when a physical function claims that it returns a type
(e.g. Int32), users can safely downcast its resulting Array to the
corresponding type (e.g. Int32Array), as well as to write data to formats that
have a schema with nullability flag (e.g. parquet).

#### Responsibility

It is the responsibility of the developer that writes a physical function to
guarantee this invariant.

In particular:

* The derived DataType matches the code it uses to build the array for every branch of valid input type combinations.
* The nullability flag matches how the values are built.

#### Validation

Since the validation of this invariant is computationally expensive, execution
contexts CAN validate this invariant.

### The physical schema is invariant under planning

The physical schema derived by a physical plan returned by the planner MUST be
equivalent to the physical schema derived by the logical plan passed to the
planner. Specifically:

```
plan(logical_plan).schema === logical_plan.physical_schema
```

> Logical plan's physical schema is defined as logical schema with relation

Reviewer: I am not sure about this -- wouldn't it be more like "logical schema with relation qualifiers squashed into field names"?

So input where `(relation: my_table, field: my_column)` --> `(my_table.my_column)` ?

Or perhaps this is saying internally `(my_table.my_column)` is the field name?

Author: Yeah, this is a little bit counter intuitive. Basically physical field names are not supposed to include relation qualifiers. Here is a more detailed write up on why this is the case at https://docs.google.com/document/d/158gbfDp8pcakfriT2l7dHChwJB43_RV7lcWfxEC73ng/edit#heading=h.vkwn4d4zskrs. In short, relations need to be decoupled from record batch outputs so we can load the same datafusion output into different relations later.

In the beginning of this spec's Notation section, we defined a physical schema as:

```
Physical schema: a vector of physical fields, used by both physical plan and Arrow record batch.
```

The goal here is to make sure the physical schema derived from the logical plan will match exactly with the physical schema of the physical plan and record batches. And since physical plan and record batch schema fields are not supposed to contain relation qualifiers, we would need to do the same thing for logical plan's physical schema as well.

If you think the context on why relation should be tracked externally outside of physical schemas is worth documenting in this doc, I can add that google doc section as appendix as well.

Reviewer: I think it is good enough as written.

```

This is used to ensure that the physical schema of its (logical) plan is what
it gets in record batches, so that users can rely on the optimized logical plan
to know the resulting physical schema.

Note that since a logical plan can be as simple as a single projection with a
single function, `Projection f(c1,c2)`, a corollary of this is that the
physical schema of every `logical function -> physical function` must be
invariant under planning.

#### Responsibility

Developers of physical and logical plans and planners MUST guarantee this
invariant for every triplet (logical plan, physical plan, conversion rule).

#### Validation

Planners MUST validate this invariant. In particular they MUST return an error
when, during planning, a physical function's derived schema does not match the
logical functions' derived schema.

### The output schema equals the physical plan schema

The schema of every RecordBatch in every partition outputted by a physical plan
MUST be equal to the schema of the physical plan. Specifically:

```
physical_plan.evaluate(batch).schema = physical_plan.schema
```

> Together with other invariances, this ensures that the consumers of record

Reviewer: ```suggestion
Together with other invariants, this ensures that the consumers of record
```
